Inboveg (#158)
inbo/tutorials: content/tutorials/r_inboveg/index.Rmd

Ample selection and export functions toward analysis tools are provided. It also
provides standardized lists of species, habitats, life forms, scales etc. 
Original observations are preserved and a full history of subsequent 
identifications is saved.

# Aim
In this tutorial we make functions available to query data directly from the 
INBOVEG SQL-server database. This to avoid writing your own queries or 
to copy/paste them from the access-frontend for INBOVEG.

We have provided functions to query

* survey (INBOVEG-projects)

* recordings (vegetation relevés)

* metadata of recordings (header info)

* classification (Natura2000 or local classification like BWK)

* qualifiers (management and site characteristics)
  
# Packages and connection
In order to run the functionalities, some R packags need to be installed. 

The main functions that we will use in this tutorial all start with `inboveg_*`. These functions are made available by loading the `inborutils` package.

You can install inborutils from github with:

```{r, eval = FALSE}
install.packages("devtools")
devtools::install_github("inbo/inborutils")
```

> This tutorial will only work for people with access to the INBO network. As an INBO employee, you should make sure you have reading-rights for CYDONIA, otherwise place an ICT-call (ict.helpdesk@inbo.be)

Reviewer: ```suggestion
This tutorial will only work for people with access to the INBO network. As an INBO employee, you should make sure you have reading-rights for CYDONIA, otherwise place an ICT-call.
```

```{r load_libraries, message = FALSE, warning = FALSE}
library(glue)
library(DBI)
library(assertthat)
library(dplyr)
library(inborutils)
```

The following R-code can be used to establish a connection to INBOVEG 
by using 'connect_inbo_dbase' of the inborutils-package with the database 'Cydonia' 
on the inbo-sql07-prd server:

```{r connection_inbo}
con <- connect_inbo_dbase("D0010_00_Cydonia")
```

# Functionality

## Survey information

The function `inboveg_survey` queries the INBOVEG database for survey information 
(metadata about surveys) for one or more survey(s) by the name of the survey.

###Examples

> Three examples are given, this can be used as base to continue selecting the wanted data

Reviewer: ```suggestion
Three examples are given, this can be used as base to continue selecting the wanted data.
```

Get information of a specific survey and collect data.

```{r examples_survey}
survey_info <- inboveg_survey(con, 
                              survey_name = "OudeLanden_1979", 
                              collect = TRUE)
```

```{r}
survey_info
```

Get information of all surveys. This time we will not use `collect = TRUE`, which will return a [lazy query](https://docs.lucee.org/guides/cookbooks/lazy_queries.html):

```{r}
allsurveys <- inboveg_survey(con)
```

```{r}
allsurveys
```

> If only a part of the survey name is known?

Reviewer: ```suggestion
If only a part of the survey name is known, you can make use of wildcards such as %.
```

This function `inboveg_qualifiers`queries the INBOVEG database for
qualifier information on recordings for one or more surveys. These qualifiers 
give information on management, location description, ... 

###Examples

Four examples are given, this can be used as base to continue selecting the wanted data.

Get the qualifiers from one survey

```{r examples_qualifiers}
qualifiers_heischraal2012 <- inboveg_qualifiers(con,
                                  survey_name = "MILKLIM_Heischraal2012")
```

```{r}
head(qualifiers_heischraal2012, 10)
```

Get all qualifiers from MILKLIM surveys (partial matching):
```{r}
qualifiers_milkim <- inboveg_qualifiers(con,
                                        survey_name = "%MILKLIM%")
```

```{r}
head(qualifiers_milkim, 10)
```

Get qualifiers from several specific surveys
```{r}
qualifiers_severalsurveys <- inboveg_qualifiers(con, 
                survey_name = c("MILKLIM_Heischraal2012", "NICHE Vlaanderen"),
                multiple = TRUE)
```

```{r}
head(qualifiers_severalsurveys, 10)
```

Get all qualifiers of all surveys

```{r}
allqualifiers <- inboveg_qualifiers(con)
```

```{r}
head(allqualifiers,10)
```

## More complex queries

> These functions gives the basis information out of INBOVEG. If more precise information is needed 'dplyr' is the magic word. In future more complex functions can be build to help the inboveg-users

Reviewer: ```suggestion
These functions give basis information out of INBOVEG.  If more detailed information is needed 'dplyr' is the magic word. In future more complex functions can be build to help the inboveg-users.
```

Natura 2000 monitoring) and the classic relevés. The classic relevés can stand 
alone, be an element of a collection or element of a chain where the linkage is 
used to give information about the relative position of recording within a series.
Ample selection and export functions toward analysis tools are provided. It also
provides standardized lists of species, habitats, life forms, scales etc. 
Original observations are preserved and a full history of subsequent 
identifications is saved.

# Aim
In this tutorial we make functions available to query data directly from the 
INBOVEG SQL-server database. This to avoid writing your own queries or 
to copy/paste them from the access-frontend for INBOVEG.

We have provided functions to query

* survey (INBOVEG-projects)

* recordings (vegetation relevés)

* metadata of recordings (header info)

* classification (Natura2000 or local classification like BWK)

* qualifiers (management and site characteristics)
  
# Packages and connection
In order to run the functionalities, some R packags need to be installed. 

The following packages are needed to run this code:

* glue

* DBI

* assertthat

> * dplyr

Author: indeed, adapted in r-studio

Loading the functionality can be done by loading the `inborutils` package:

* inborutils

You can install inborutils from github with:
install.packages("devtools")

> devtools::install_github("inbo/inborutils")

Author: done in r-studio

Be sure you have reading-rights for CYDONIA
otherwise place an ICT-call (ict.helpdesk@inbo.be)

```{r load_libraries, message = FALSE, warning = FALSE}
library(glue)
library(DBI)
library(assertthat)
library(dplyr)
library(inborutils)
```

The following R-code can be used to establish a connection to INBOVEG 
by means of a connection string:

<!--better to use a connection string than dsn. 
dsn requires extra steps and settings in windows odbc manager-->

```{r connection_str}
connection <- dbConnect(odbc::odbc(),
.connection_string = "Driver=SQL Server;Server=inbo-sql07-prd.inbo.be,1433;
Database=D0010_00_Cydonia;Trusted_Connection=Yes;")
```

Or using dbconnection of the inborutils-package with the database 'Cydonia' 

> on the inbo-sql07-prd server:

Author: indeed, less complex and only INBO-members can have access to inboveg-database

###Examples

Four examples are given, this can be used as base to continue selecting the wanted data

Get the qualifiers from one survey

```{r examples_qualifiers}
qualifiers_heischraal2012 <- inboveg_qualifiers(con,
                                  survey_name = "MILKLIM_Heischraal2012")
```

```{r}
head(qualifiers_heischraal2012, 10)
```

get all qualifiers from MILKLIM surveys (partial matching)
```{r}
qualifiers_milkim <- inboveg_qualifiers(con,
                                        survey_name = "%MILKLIM%")
```

```{r}
head(qualifiers_milkim, 10)
```

Get qualifiers from several specific surveys
```{r}
qualifiers_severalsurveys <- inboveg_qualifiers(con, 
                survey_name = c("MILKLIM_Heischraal2012", "NICHE Vlaanderen"),
                multiple = TRUE)
```

```{r}
head(qualifiers_severalsurveys, 10)
```

Get all qualifiers of all surveys

```{r}
allqualifiers <- inboveg_qualifiers(con)
```

```{r}
head(allqualifiers,10)
```

## More complex queries

These functions gives the basis information out of INBOVEG. If more precise information is needed 'dplyr' is the magic word.

### Examples

Still in progress

> 

Author: They are indeed of no use at this stage. But actually it is a kind of reminder for my fuzzy brain to extend this tutorial with more complex cases later on (and if asked by the inboveg-users.

> 

Author: rewrote the chapter 'More complex queries, deleted 'Examples'
